### PR TITLE
Add Clojure language support with tree-sitter and LSP

### PR DESCRIPTION
**Pull request by prcastro**
Current limitations:
* Not able to navigate into JAR files 

Release Notes:

- Added Clojure language support

### CODE REVIEW

**Comment by prcastro** [comment]
@cla-bot check

**Comment by prcastro** [comment]
Hey! I'd like some help with this PR. Currently the LSP is being downloaded and extracted correctly (at least it seems) but nothing LSP related it running on the editor.

How can I see the LSP logs within Zed?

**Comment by pseudomata** [comment]
@prcastro you can view LSP logs by hitting `cmd+shift+p` and typing "debug: open language server logs". There are also logs in zed "zed: open log".

File: crates/zed/src/languages/clojure/config.toml
Comment (ericdallo) [inline crates/zed/src/languages/clojure/config.toml:2]: WDYT about adding `edn`, `bb`, `cljd` extensions too?

File: crates/zed/src/languages/clojure.rs
Comment (ericdallo) [inline crates/zed/src/languages/clojure.rs:41]: is that hard to support linux as well? it should follow the same standard

Comment (prcastro) [reply crates/zed/src/languages/clojure.rs:41]: Zed is macos only for now

Comment (pseudomata) [reply crates/zed/src/languages/clojure.rs:41]: Zed is macOS only for now (see https://github.com/zed-industries/zed/issues/5395)

**Comment by prcastro** [comment]
Running "debug: open language server logs" makes Zed hang =s I'm checking with Eric how I can access the logs manually.

**Comment by prcastro** [comment]
@pseudomata I'm getting something like this:

```
[2024-01-29T18:28:03Z ERROR lsp] Cancelled LSP request task for "initialize" id 0 which took over 120s
```

Talking to @ericdallo we reached the conclusion the client is timing out before the LSP finishes opening the project (which includes even downloading the dependencies, afaik). Is there a way to increate the timeout of 120s?

**Comment by prcastro** [comment]
It doesn't take long for small projects usually. However, it took a long time (more than 2min) to open a big project with a lot of dependencies to download and compile. About GraalVM, I'm not sure, maybe @ericdallo can respond.

Also I made it work. Here are some limitations of my approach:
* I had to increase the timeout to make it work (otherwise the LSP might not have enough time to start)
* I currently cannot navigate into variables defined in JARs, because I receive the following error:

```
[2024-01-29T18:59:56Z ERROR project::worktree] failed to canonicalize root path: No such file or directory (os error 2)
[2024-01-29T18:59:56Z ERROR util] crates/editor/src/editor.rs:7045: ProjectPath { worktree_id: WorktreeId(30064771211), path: "" } opening failure: No such file or directory (os error 2)
```

For me those are not huge problems. WDYT?

File: crates/lsp/src/lsp.rs
Comment (prcastro) [inline crates/lsp/src/lsp.rs:36]: Had to do this to give enough time for clojure-lsp to initialize

Comment (osiewicz) [reply crates/lsp/src/lsp.rs:36]: Why does closure-lsp take that long in the first place though? Is that something that e.g. other IDEs using this LSP server suffer from as well?

**Comment by SomeoneToIgnore** [comment]
This looks close to be done, so it would be great to add a new docs entry to the https://github.com/zed-industries/zed/tree/main/docs/src/languages list before merging.

File: crates/zed/src/languages/clojure/config.toml
Comment (osiewicz) [inline crates/zed/src/languages/clojure/config.toml:3]:
```suggestion
line_comments = ["# "]
```
We've changed the config for `line_comments` to array recently.

File: crates/lsp/src/lsp.rs
Comment (prcastro) [reply crates/lsp/src/lsp.rs:36]: Most other editors don't set a timeout as far as I know. And yes, I was running the project in an "extreme" setting: my project has a ton of dependencies and I'm behind a VPN so downloading is a bit slow. Also, the project takes a while to compile.

But let me say I work at Nubank, which might be the biggest Clojure shop in the world, and most of our projects are similar to this. So I'd say it's pretty important for Clojure users.

Comment (prcastro) [reply crates/lsp/src/lsp.rs:36]: And yes, I have to wait a while for the lsp when opening a big project for the first time

Comment (osiewicz) [reply crates/lsp/src/lsp.rs:36]: I mean, I do have to wait for rust-analyzer as well (see the attached screenshot), but it still sends a response to initialize request within 28ms. That's kind of my point; regardless of your project size, response to `Initialize` should ideally be pretty snappy.
image
All I'm saying is that I don't think we should be bumping that timeout to 2 minutes just to accommodate for what (to me) looks like an anomaly rather than a norm. If you could point me to a pretty big open-source Closure project (whose size is somewhat representative of what you're dealing with) I could profile the `closure-lsp` with, that'd be golden.

Comment (osiewicz) [reply crates/lsp/src/lsp.rs:36]: @prcastro when you run the build from this branch with `RUST_LOG=lsp=trace`, is there anything that stands out from the logs between us sending the initialise request and receiving a response to that request?
For example, here's my log for `RUST_LOG=lsp=trace cargo run -- ../random_projects/clojure_lsp`:
<details><summary>Details</summary>
<p>

[2024-01-29T21:46:11Z TRACE lsp] outgoing message:{"jsonrpc":"2.0","id":0,"method":"initialize","params":{"processId":null,"rootUri":"file:///","capabilities":{"workspace":{"workspaceEdit":{"documentChanges":true,"resourceOperations":["create","rename","delete"]},"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true,"relativePatternSupport":true},"symbol":{},"workspaceFolders":true,"configuration":true,"inlayHint":{"refreshSupport":true},"diagnostic":{}},"textDocument":{"completion":{"completionItem":{"snippetSupport":true,"resolveSupport":{"properties":["additionalTextEdits"]}},"completionList":{"itemDefaults":["commitCharacters","editRange","insertTextMode","data"]}},"hover":{"contentFormat":["markdown"]},"formatting":{},"onTypeFormatting":{},"definition":{"linkSupport":true},"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["refactor","quickfix","source"]}},"dataSupport":true,"resolveSupport":{"properties":["edit","command"]}},"rename":{"prepareSupport":true},"publishDiagnostics":{"relatedInformation":true},"inlayHint":{"dynamicRegistration":false,"resolveSupport":{"properties":["textEdits","tooltip","label.tooltip","label.location","label.command"]}},"diagnostic":{"relatedDocumentSupport":true}},"window":{"workDoneProgress":true},"experimental":{"serverStatusNotification":true}},"workspaceFolders":[{"uri":"file:///","name":""}]}}
[2024-01-29T21:46:11Z TRACE lsp] incoming message: {"jsonrpc":"2.0","method":"LogMessage","params":{"level":0,"message":"[DEBUG] [agent] [2024-01-29T21:46:11.717Z] Agent service starting","metadataStr":"[DEBUG] [agent] [2024-01-29T21:46:11.717Z]","extra":["Agent service starting"]}}
[2024-01-29T21:46:11Z INFO  lsp] 0 unhandled notification LogMessage:
    {
      "level": 0,
      "message": "[DEBUG] [agent] [2024-01-29T21:46:11.717Z] Agent service starting",
      "metadataStr": "[DEBUG] [agent] [2024-01-29T21:46:11.717Z]",
      "extra": [
        "Agent service starting"
      ]
    }
[2024-01-29T21:46:11Z TRACE lsp] incoming message: {"jsonrpc":"2.0","id":0,"method":"client/registerCapability","params":{"registrations":[{"id":"1148dea4-b570-478c-bd60-8a41da3242dc","method":"workspace/didChangeWorkspaceFolders","registerOptions":{}}]}}
[2024-01-29T21:46:11Z INFO  lsp] 0 unhandled notification client/registerCapability:
    {
      "registrations": [
        {
          "id": "1148dea4-b570-478c-bd60-8a41da3242dc",
          "method": "workspace/didChangeWorkspaceFolders",
          "registerOptions": {}
        }
      ]
    }
[2024-01-29T21:46:11Z TRACE lsp] incoming message: {"jsonrpc":"2.0","id":0,"result":{"capabilities":{"textDocumentSync":{"openClose":true,"change":2},"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true}}}}}
[2024-01-29T21:46:11Z TRACE lsp] Took 240.186334ms to receive response to "initialize" id 0

</p>
</details> 
You do not need to share the logs as I assume they might contain sensitive data, but maybe clojure-lsp has logs that could help us out here.

Comment (prcastro) [reply crates/lsp/src/lsp.rs:36]: @osiewicz I'm willing to revert the timeout to 2 minutes and then, on another pull request, make that configurable by the user. What do you think?

BTW, maybe you can test with this project: https://github.com/riemann/riemann

**Comment by prcastro** [comment]
> This looks close to be done, so it would be great to add a new docs entry to the https://github.com/zed-industries/zed/tree/main/docs/src/languages list before merging.

Done!

File: crates/lsp/src/lsp.rs
Comment (osiewicz) [reply crates/lsp/src/lsp.rs:36]: > I'm willing to revert the timeout to 2 minutes and then, on another pull request, make that configurable by the user. What do you think?

I think this is a great idea, let's do that. Still, I think we should also get to the bottom of this and find a reason for a slow response to initialize. :p

It takes 200ms on my machine to get response to initialize request on that riemann repo you've shared.

Comment (ericdallo) [reply crates/lsp/src/lsp.rs:36]: clojure-lsp really could take 1-2mins or more depending on the size of the project to respond to initialize request, it's on that request that analyze and scan the classpath and build its ASM, vscode, emacs and vim don't have timeouts IIRC on that request, I think would be nice to do the same as the LSP spec doesn't mention that

Comment (osiewicz) [reply crates/lsp/src/lsp.rs:36]: Right, let's just add the knob to configure a timeout for initialize and call it a day. =)

Comment (prcastro) [reply crates/lsp/src/lsp.rs]: I reverted the timeout on this PR, and I'll soon start working on this configuration knob.

**Comment by prcastro** [comment]
@osiewicz do you mind taking another look?

File: Cargo.toml
Comment (osiewicz) [inline Cargo.toml:137]:
```suggestion

```

**Comment by osiewicz** [comment]
Yup, LGTM on the first skim. I'll give it a go on some clojure project later.

File: Cargo.toml
Comment (prcastro) [reply Cargo.toml]: Sorry about that, happened when I was merging with the commits on master

Comment (osiewicz) [reply Cargo.toml]: No worries ^.^

**Comment by osiewicz** [comment]
It looks like VSC does not have the problems with initialisation request (it completes within a reasonable time frame there), so there's definitely something Zed can do better there. I'm gonna try to work the differences out tomorrow.
I am using Calva extension there and inspecting solely it's communication with clojure-lsp.

File: crates/zed/src/languages/clojure/config.toml
Comment (osiewicz) [inline crates/zed/src/languages/clojure/config.toml:10]: We must also specify `word_characters = ["-"]`, as that has significant impact on how navigating around source files in a given language works. E.g. without it "go to word end" when your cursor is at the start of `some-identifier` will go to `-`, not `r`.

**Comment by osiewicz** [comment]
One issue I've ran into is that clojure-lsp sends a request to the client to show a popup when it encounters issues during initialisation. e.g:
```
[Trace - 00:12:46] Received request 'window/showMessageRequest - (1)'.
Params: {
    "type": 2,
    "message": "LSP classpath lookup failed when running `lein with-profile +test,+dev classpath`. Some features may not work properly if ignored.\n\nError: Cannot run program \"lein\" (in directory \"/Users/hiro/Projects/throwaways/riemann\"): error=2, No such file or directory\n\nChoose an option:",
    "actions": [
        {
            "title": "Retry"
        },
        {
            "title": "Ignore"
        }
    ]
}


[Trace - 00:12:49] Sending response 'window/showMessageRequest - (1)'. Processing request took 3332ms
Result: {
    "title": "Ignore"
}
````
I don't have `lein` installed, hence the popup in Vscode. Funnily enough, response to initialize request comes right after that, so I suspect it's blocked on that response.
We do not handle that request, hence clojure-lsp waits around forever.
I don't believe we have API to display generic messages to user (in a way that they can choose an option), so we'd have to do a bit of work to handle the request proper. Alternatively, we could require that lein is installed on user machine (which we can do in `fetch_server_binary` or whatever). 
Opening Riemann in VSCode does indeed take 2 minutes to complete initialize request. Bummer.

**Comment by ericdallo** [comment]
@osiewicz correct, the problem is that is not always lein, it depends on the type of clojure project, it could be `lein`, `clojure`, `boot`, `babashka`, `npx`, so not sure it's a good idea to require all those, the ideal would be to prompt to user

**Comment by osiewicz** [comment]
Oh, I see; yeah, then we definitely need to prompt the user for an answer.

**Comment by lanjoni** [comment]
At this point is it possible to evaluate expressions using interactive REPL? Similar to `cider-jack-in` using Emacs...

If not, how can this be implemented?

**Comment by prcastro** [comment]
Hey @osiewicz , just to understand the steps we need to take to get this merged:

- Add dash to word characters
- Implement a modal to send messages that require inputs to users
- Implement a handler for the showMessageRequest that presents the modal to the user

Is that it? Any pointers to get started with step 2? I'll probably need a tip on how to get that started.

**Comment by osiewicz** [comment]
Hey, 
I'd be down to pair on that sometime this week (preferably on Thursday) if needed.
To display a modal, we could use something like `WindowContext::prompt`, which isn't really ideal (it'll look like the prompt you get for unsaved files, which I'm not sure if that's ideal); I think it's fine to start with that, we can touch up the prompts at some point down the line.
For a `showMessageRequest` handler, I'd recommend using `Workspace::show_notification` - UI for that is in a good shape I believe, that will look like a update notification

**Comment by prcastro** [comment]
@osiewicz I would appreciate pairing on this issue on Thursday if you are available. Can you add me on Twitter so we can DM the details? It's @poliveiracastro over there

**Comment by osiewicz** [comment]
Sure, I've just followed & dmed you

**Comment by prcastro** [comment]
@osiewicz can you check your inbox?

**Comment by prcastro** [comment]
Hey @osiewicz, I just finish what we started earlier today. Can you take a look and see if this PR is in an acceptable state now? Just to recap, we have some limitations:

- Zed doesn't install lein for you when it's installing the LSP
- We don't support navigating into a symbol (e.g. a function) defined in a JAR
- We handle `showMessageRequest` by showing a `WindowContext::prompt` which is a  macOS native prompt (not a prompt made in GPUI or anything). 
- We will still fail if the LSP takes too long to reply

File: crates/project/src/project.rs
Comment (osiewicz) [inline crates/project/src/project.rs:219]: Don't we need to use that prompt level in workspace.rs? e.g. here: https://github.com/zed-industries/zed/pull/6988/files#diff-041673bbd1947a35d45945636c0055429dfc8b5985faf93f8a8a960c9ad31e28R585

**Comment by osiewicz** [comment]
Yeah, I think that's fair; I think these can be fleshed out in a follow-up PR, this one relatively self-contained.

**Comment by prcastro** [comment]
@osiewicz Any idea why this is happening?

```
error[E0061]: this function takes 4 arguments but 3 arguments were supplied
  --> crates/zed/src/languages/clojure.rs:30:13
   |
30 |             latest_github_release("clojure-lsp/clojure-lsp", false, delegate.http_client()).await?;
   |             ^^^^^^^^^^^^^^^^^^^^^                                   ---------------------- an argument of type `bool` is missing
```

I could swear that this function call is correct

**Comment by osiewicz** [comment]
I think that's due to  https://github.com/zed-industries/zed/pull/7571/files#diff-c9e13c01768f22f905b924b7fc003f8288d1f6ab3d1b6f5232e178ef0162cd12 - IIRC merges are tested against main (or on top of main should I say), so once you merge main into your branch, the code should stop compiling.

**Review by osiewicz — APPROVED** [review]
LGTM, thanks @prcastro :)

**Comment by maxbrunsfeld** [comment]
Nice work, @prcastro. 

Just a heads up - we are very soon going to be moving most of Zed's built-in languages into installable extensions (https://github.com/zed-industries/zed/issues/7096). Clojure is a popular enough language that I'm ok with merging this in the meantime. But once the extension system lands, we will probably remove this code and put it into a new `clojure` repository under the new [zed-extensions](https://github.com/orgs/zed-extensions/repositories) GitHub org. We'll probably add you as a collaborator on that repo when we do it, if you're interested in helping maintain it.

**Comment by aviflax** [comment]
Thank you @prcastro and @osiewicz for your work on this! I’m using it and enjoying it. 🍻

**Pull request merged**
